### ISSUE TITLE
M2.2.3 : Unexpected behaviour when programmatically assigning products to a category

### ISSUE THREAD

**Issue by kanduvisla** [magento/magento2#14861]
I'm trying to add Category Links to a product using the extension attributes-way, but instead of replacing the categories, they get added.

### Preconditions

- Magento 2.2.3
- Create a bunch of categories with ID 1 to 10

### Steps to reproduce

I wrote a code example:

```
/**
 * @param \Magento\Catalog\Model\Product $product
 * @param array $categories
 */
function saveCategories($product, $categories) {
    $categoryLinks = [];
    
    foreach ($categories as $categoryData) {
        /** @var CategoryLinkInterface $categoryLink */
        $categoryLink = $this->categoryLinkFactory->create();
        $categoryLink->setCategoryId($categoryData['category_id']);
        $categoryLink->setPosition($categoryData['position']);
        $categoryLinks[] = $categoryLink;
    }

    $extensionAttributes = $product->getExtensionAttributes();
    $extensionAttributes->setCategoryLinks($categoryLinks);
    $product->setExtensionAttributes($extensionAttributes);
    $this->productRepository->save($product);
}

// Start example code:

$categories = [
    ['category_id' => 3, 'position' => 0,],
    ['category_id' => 4, 'position' => 0,],
    ['category_id' => 5, 'position' => 0,],
];

saveCategories($product, $categories);

// Check:

$product = $this->productRepository->load($product->getSku());
var_dump($product->getCategoryIds()); // [3, 4, 5]

// Again:

$categories = [
    ['category_id' => 6, 'position' => 0,],
    ['category_id' => 7, 'position' => 0,],
];

saveCategories($product, $categories);

// Check:

$product = $this->productRepository->load($product->getSku());
var_dump($product->getCategoryIds()); // [3, 4, 5] = WRONG!
```

### Expected result

The last result should return ID's `[6, 7]`, since those where the only category links set.

### Actual result

The last result returns ID's `[3, 4, 5]`, which indicates nothing has changed.

### Possible cause?

If I look at `\Magento\Catalog\Model\ResourceModel\Product\CategoryLink::saveCategoryLinks()` the logic seems to be a bit off. The `$delete` and `$insert`-variables don't look properly formed. For example: the `$delete`-variable is always empty, whilst I would expect it to be populated if I change for ID `3,4,5` to `6,7`. The `$deleteUpdate`-variable is also empty. So could the problem be in `\Magento\Catalog\Model\ResourceModel\Product\CategoryLink::processCategoryLinks()`?

**Comment by kanduvisla** [comment]
On inspecting, I think the problem can be found in `Magento\Catalog\Model\ResourceModel\Product\CategoryLink::saveCategoryLinks()` and it's use of the `processCategoryLink()`-method. The second argument is passed as a reference on this method and is done so to be able to `unset()` some keys:

    /**
     * Process category links
     *
     * @param array $newCategoryPositions
     * @param array $oldCategoryPositions
     * @return array
     */
    private function processCategoryLinks($newCategoryPositions, &$oldCategoryPositions)
    {
        $result = ['changed' => [], 'updated' => []];
        foreach ($newCategoryPositions as $newCategoryPosition) {
            $key = array_search(
                $newCategoryPosition['category_id'],
                array_column($oldCategoryPositions, 'category_id')
            );

            if ($key === false) {
                $result['changed'][] = $newCategoryPosition;
            } elseif ($oldCategoryPositions[$key]['position'] != $newCategoryPosition['position']) {
                $result['updated'][] = $newCategoryPositions[$key];
                unset($oldCategoryPositions[$key]);
            }
        }

        return $result;
    }

And this method is called like this:

    $insertUpdate = $this->processCategoryLinks($categoryLinks, $oldCategoryLinks);
    $deleteUpdate = $this->processCategoryLinks($oldCategoryLinks, $categoryLinks);

So let's see what happens, step by step: In my situation I have the following variables:

```
$categoryLinks = [
    0 => [
        'category_id' => '3',
        'position' => '0',
    ],
    1 => [
        'position' => 0,
        'category_id' => '6',
    ],
    2 => [
        'position' => 0,
        'category_id' => '7',
    ],
    3 => [
        'category_id' => '4',
        'position' => '0',
    ],
    4 => [
        'category_id' => '5',
        'position' => '0',
    ],
];

$oldCategoryLinks = [
    0 => [
        'category_id' => '3',
        'position' => '0',
    ],
    1 => [
        'category_id' => '4',
        'position' => '0',
    ],
    2 => [
        'category_id' => '5',
        'position' => '0',
    ],
];
```

The `$insertUpdate`-parameter is always correctly filled; it has the ID's that I expect. The `$deleteUpdate` however is empty in both cases. I would expect this one to be populated with the ID's of the old categories on the second run, because those are the links that ought to be deleted.

Can anyone take a look at this? Or see if they can reproduce this? I can't seem to find the cause of the issue.

**Comment by kanduvisla** [comment]
On a sidenote: I also seem to be unable to change a single category using this way. Or am I just misunderstanding how to use CategoryLinks?

For example: 

- I create a product with 1 category link with category ID = 3.
- Next up I load that same product but replace the category links with **one** new category with ID = 4.
- Now the product is assigned to 2 categories: 3 & 4, instead of only 4.

To make it even more interesting, even if I flush the table priorly I get the error :-|

    $connection = $this->resourceConnection->getConnection();            
    $connection->delete(
        'catalog_category_product',
        $connection->quoteInto('product_id = ?', $product->getId())
    );

I even tried with caching disabled!

**Comment by kanduvisla** [comment]
I found a workaround for this issue: If I always set an empty array for the 'native' `setCategoryIds()`-method, it works as expected:

```
$product->setCategoryIds([]);
$extensionAttributes = $product->getExtensionAttributes();
$extensionAttributes->setCategoryLinks($categoryLinks); // = not empty
$product->setExtensionAttributes($extensionAttributes);
$this->productRepository->save($product);
```

So it seems that the 'extension attribute'-approach and the old-skool `setCategoryIds()` / `getCategoryIds()` are biting each other somehow.

**Comment by engcom-backlog-nickolas** [comment]
Hello @kanduvisla, thank you for your report.
We've acknowledged the issue and added to our backlog.

**Comment by kanduvisla** [comment]
I've bumped into this issue once again this day. I have a dataset that can replicate this issue:

```
$newCategoryPositions = [
    0 =>
        [
            'position' => '1',
            'category_id' => '28',
        ],
    1 =>
        [
            'position' => '2',
            'category_id' => '36',
        ],
];

$oldCategoryPositions = [
    0 =>
        [
            'category_id' => '28',
            'position' => '2',
        ],
    1 =>
        [
            'category_id' => '36',
            'position' => '1',
        ],
];

```

If you run `\Magento\Catalog\Model\ResourceModel\Product\CategoryLink::processCategoryLinks()` with above dataset, it will trigger an "undefined offset" error. Even though it's perfectly valid to swap the positions of 2 category links.

**Comment by kanduvisla** [comment]
The error can be reproduced by adding the following to `\Magento\Catalog\Test\Unit\Model\ResourceModel\Product\CategoryLinkTest::getCategoryLinksDataProvider()`:

```
            [
                [
                    ['category_id' => 3, 'position' => 10],
                    ['category_id' => 4, 'position' => 20],
                ],
                [
                    ['category_id' => 3, 'position' => 20], // swapped positions
                    ['category_id' => 4, 'position' => 10], // swapped positions
                ],
                [3, 4] // 3, 4 - updated
            ]
```

This test is weak by the way, it only checks which ID's are affected, but not if the proper insert, update or delete actions are executed. I'll see if I can improve this test and fix this issue while at it.

**Comment by magento-engcom-team** [comment]
Hi @kanduvisla. Thank you for your report. 
The issue has been fixed in magento/magento2#19080 by @kanduvisla in 2.3-develop branch
Related commit(s):
 - [39055807e508b33d658f18e77caec732b9aaca2b](https://github.com/magento/magento2/commit/39055807e508b33d658f18e77caec732b9aaca2b)
 - [c2100db5bd2ec177f0fdf5a40ff9ac97e46c0b1a](https://github.com/magento/magento2/commit/c2100db5bd2ec177f0fdf5a40ff9ac97e46c0b1a)
 - [d904697ec173a2423750079ba1d2b3d76f62dcf5](https://github.com/magento/magento2/commit/d904697ec173a2423750079ba1d2b3d76f62dcf5)
 - [9b919f93479c68dba0a4c8a0936a7f408438372b](https://github.com/magento/magento2/commit/9b919f93479c68dba0a4c8a0936a7f408438372b)
 - [9cb7a895045ff591f27b66a06cbb575d7aa62bc9](https://github.com/magento/magento2/commit/9cb7a895045ff591f27b66a06cbb575d7aa62bc9)

The fix will be available with the upcoming 2.3.1 release.

**Comment by Tomasz-Silpion** [comment]
If anyone has experienced the similar issue remember to assign the product to categories with CategoryLinkManagementInterface AFTER the product was saved. That's how it was done in \Magento\Catalog\Controller\Adminhtml\Product\Save action.

```
$product->save();
$this->categoryLinkManagement->assignProductToCategories(
     $product->getSku(),
     $categoryIds
); 
```
